File: gateway/src/main/java/com/beat/gateway/guest/internal/GuestSessionService.java
package com.beat.gateway.guest.internal;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.OptionalLong;

import com.beat.contracts.auth.guest.GuestSessionPort;
import com.beat.gateway.guest.internal.store.GuestSession;
import com.beat.gateway.guest.internal.store.GuestSessionRepository;

public final class GuestSessionService implements GuestSessionPort {

	private static final int TOKEN_BYTE_LENGTH = 32;

	private final GuestSessionRepository guestSessionRepository;
	private final SecureRandom secureRandom = new SecureRandom();

	public GuestSessionService(GuestSessionRepository guestSessionRepository) {
		this.guestSessionRepository = guestSessionRepository;
	}

	@Override
	public String issue(long userId) {
		byte[] tokenBytes = new byte[TOKEN_BYTE_LENGTH];
		secureRandom.nextBytes(tokenBytes);
		String token = Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes);
		guestSessionRepository.save(GuestSession.of(hash(token), userId));
		return token;
	}

	@Override
	public OptionalLong findUserId(String token) {
		if (token == null || token.isBlank()) {
			return OptionalLong.empty();
		}
		return guestSessionRepository.findById(hash(token))
			.map(GuestSession::getUserId)
			.map(OptionalLong::of)
			.orElseGet(OptionalLong::empty);
	}

	private String hash(String token) {
		try {
			byte[] digest = MessageDigest.getInstance("SHA-256")
				.digest(token.getBytes(StandardCharsets.UTF_8));
			return Base64.getUrlEncoder().withoutPadding().encodeToString(digest);
		} catch (NoSuchAlgorithmException exception) {
			throw new IllegalStateException("SHA-256 is unavailable", exception);
		}
	}
}

File: gateway/src/main/kotlin/com/beat/gateway/guest/internal/GuestSessionService.kt
package com.beat.gateway.guest.internal

import com.beat.contracts.auth.guest.GuestSessionPort
import com.beat.gateway.guest.internal.store.GuestSession
import com.beat.gateway.guest.internal.store.GuestSessionRepository
import java.security.SecureRandom
import java.util.Base64
import java.util.OptionalLong

class GuestSessionService(
    private val guestSessionRepository: GuestSessionRepository,
) : GuestSessionPort {

    private val secureRandom = SecureRandom()

    override fun issue(userId: Long): String {
        val token = generateToken()
        guestSessionRepository.save(GuestSession(Sha256Hasher.hashToBase64Url(token), userId))
        return token
    }

    override fun findUserId(token: String): OptionalLong {
        if (token.isBlank()) {
            return OptionalLong.empty()
        }
        return guestSessionRepository.findById(Sha256Hasher.hashToBase64Url(token))
            .map { OptionalLong.of(it.userId) }
            .orElseGet { OptionalLong.empty() }
    }

    private fun generateToken(): String {
        val tokenBytes = ByteArray(TOKEN_BYTE_LENGTH).also(secureRandom::nextBytes)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes)
    }

    companion object {
        private const val TOKEN_BYTE_LENGTH = 32
    }
}
